Source/Csla.Data.EF7.Shared/DbContextManager.cs

```diff
@@ -0,0 +1,170 @@
+//-----------------------------------------------------------------------
+#if !MONO
+using System;
+using Microsoft.Data.Entity;
+
+namespace Csla.Data.EF7
+{
+  /// <summary>
+  /// Provides an automated way to reuse 
+  /// Entity Framework DbContext objects 
+  /// within the context
+  /// of a single data portal operation.
+  /// </summary>
+  /// <typeparam name="C">
+  /// Type of database 
+  /// DbContext object to use.
+  /// </typeparam>
+  /// <remarks>
+  /// This type stores the object context object 
+  /// in <see cref="Csla.ApplicationContext.LocalContext" />
+  /// and uses reference counting through
+  /// <see cref="IDisposable" /> to keep the data context object
+  /// open for reuse by child objects, and to automatically
+  /// dispose the object when the last consumer
+  /// has called Dispose."
+  /// </remarks>
+  public class DbContextManager<C> : IDisposable where C : DbContext
+  {
+    private static object _lock = new object();
+    private C _context;
+    private string _label;
+    private string ContextLabel { get; set; }
+
+    /// <summary>
+    /// Gets the ObjectContextManager object for the     /// specified database.
+    /// </summary>
+
+    public static DbContextManager<C> GetManager()
+    {
+      return GetManager(string.Empty);
+    }
+
+    /// <summary>
+    /// Gets the ObjectContextManager object for the 
+    /// specified database.
+    /// </summary>
+    /// <param name="database">Database name as shown in the config file.</param>
+    public static DbContextManager<C> GetManager(string database)
+    {
+      return GetManager(database, "default");
+    }
+   
+    /// <summary>
+    /// Gets the ObjectContextManager object for the 
+    /// specified database.
+    /// </summary>
+    /// <param name="database">
+    /// The database name or connection string.
+    /// </param>
+    /// <param name="label">Label for this context.</param>    
+    /// <returns>ContextManager object for the name.</returns>
+    public static DbContextManager<C> GetManager(string database, string label)
+    {
+      lock (_lock)
+      {
+        var contextLabel = GetContextName(database, label);
+        DbContextManager<C> mgr = null;
+        if (ApplicationContext.LocalContext.Contains(contextLabel))
+        {
+          mgr = (DbContextManager<C>)(ApplicationContext.LocalContext[contextLabel]);
+        }
+        else
+        {
+          mgr = new DbContextManager<C>(database, label);
+          mgr.ContextLabel = contextLabel;
+          ApplicationContext.LocalContext[contextLabel] = mgr;
+        }
+        mgr.AddRef();
+        return mgr;
+      }
+    }      
+
+    private DbContextManager(string database, string label)
+    {
+      _label = label;                
+      if (string.IsNullOrEmpty(database))
+      {
+        _context = (C) (Activator.CreateInstance(typeof (C)));
+      }
+      else
+      {
+        _context = (C)(Activator.CreateInstance(typeof(C), database));
```

Comment:
That's true I guess the potential is there for that.

Part of the issue is that the arguments passed to GetManager need to be enough to create a `DbContext` instance, but as an instance is not always created (if it's being reused) then sometimes these arguments will ultimately be redundant (beyond determining the cache key).

The constructor in EF 7 for `DbContext` takes an object which cannot really be used for determining the cache key, and so it is going to be purely redundant if it its passed in to `GetManager` but not used as a `DbContext` is resolved from the cache.

So really there is a kind of "instance management" issue here with the constructor args.

I thought about:

```
DbContextManager<MyDbContext>.GetManager(options);
```

Where `options` is some instance of `DbContextOptions` that the consumer is responsible for creating.

But again - there is potential for the consumer to create lot's of redundant instances of this dotted around their code base, because they create it on demand:

```
var optionsBuilder = new DbContextOptionsBuilder();
var options = optionsBuilder.UseSqlServer("some connection string here") 
DbContextManager<MyDbContext>.GetManager(options);
```

So I kind of ruled that out.

The thing about the lamda approach is that atleast it's only invoked when the constructor arguements are actually required, because it's when a new `DbContext` is being constructed - so it kind of solves the problem of creating redundant instances of arguements that wont be used.

How about this idea then..

We define a `Factory` interface, something like 

```
public interface IDbContextOptionsFactory
{
    DbContextOptions GetOptions();
}

```

The consumer would have to implement that - by creating something like:

```
public class MyDbContextOptionsFactory : IDbContextOptionsFactory
{
    DbContextOptions GetOptions() 
 {
    var optionsBuilder = new DbContextOptionsBuilder();
    var options = optionsBuilder.UseSqlServer("some connection string here") 
    return options;
  }
}
```

Then DbContextManager usage becomes:

 ```
var optionsFactory = new MyDbContextOptionsFactory();
`DbContextManager<MyContext>.GetManager(optionsFactory)` 
```

`DbContextManager` would only invoke the factory when it needs the options because its creating a new `DbContext` instance.

That's still potentially redundant instances of the factory class being created this time though..

Now I am fresh out of ideas.. will sleep on it..